### PR TITLE
feat(memory): add TaskTypeDetector service (#142)

### LINKED ISSUE #142: feat(memory): 3B — TaskTypeDetector implementation
## Parent Epic
#139

## Phase
Phase 2 of 4 — TaskTypeDetector Implementation

## Depends on
3A (domain types)

## Problem

There is no implementation for detecting task type from user prompts. The domain interfaces define the contract (`ITaskTypeDetector`) but there is no concrete service or DI registration.

### Create `src/lib/infrastructure/services/TaskTypeDetector.ts`

Signal-based scoring implementation:

- Normalize prompt (lowercase, strip punctuation)
- Score each task type by counting keyword/phrase matches from `DETECTION_SIGNALS`
- Multi-word phrases (e.g. "what if", "not working") scored higher than single keywords
- Return highest-scoring type with confidence (0.0-1.0)
- Default to `'execution'` when no clear signal (confidence < threshold)
- Confidence = winning score / total possible signals, capped at 1.0

### Modify `src/lib/infrastructure/di/tokens.ts`

Add `TaskTypeDetector` token to `INFRA_TOKENS`.

### Modify `src/lib/infrastructure/di/bootstrap.ts`

Register `TaskTypeDetector` as transient (stateless, no dependencies).

### Export from `src/lib/infrastructure/services/index.ts`

Add `TaskTypeDetector` export.

## Files to modify
- `src/lib/infrastructure/services/TaskTypeDetector.ts` (create)
- `src/lib/infrastructure/di/tokens.ts` (modify)
- `src/lib/infrastructure/di/bootstrap.ts` (modify)
- `src/lib/infrastructure/services/index.ts` (modify)
- `tests/unit/src/lib/infrastructure/services/TaskTypeDetector.test.ts` (create)

## Acceptance Criteria

- `TaskTypeDetector.detect()` returns correct type for clear signals (e.g. "fix the bug" → debugging)
- Multi-word phrases scored appropriately ("what if we" → exploration)
- Defaults to `'execution'` for ambiguous or empty prompts
- Confidence score between 0.0 and 1.0
- `signals` array in result contains matched keywords
- Registered in DI container and resolvable
- Comprehensive tests: planning/execution/exploration/debugging prompts, edge cases, ambiguous inputs
- `npm run lint` and `npm test` pass

### PR DESCRIPTION
**Pull request by TonyCasey**
## Summary
- Implement `TaskTypeDetector` service with signal-based keyword/phrase scoring
- Multi-word phrases (e.g. "what if", "not working") weighted 2x for specificity
- Defaults to `execution` when confidence below threshold (0.15)
- Add `TaskTypeDetector` DI token and register as transient in bootstrap
- Export from infrastructure services barrel
- Add 27 unit tests covering all 4 task types, edge cases, and confidence bounds

Closes #142

## Test plan
- [x] `npm run lint` passes on all changed files
- [x] `npx tsc --noEmit` passes
- [x] 27/27 new tests pass (`TaskTypeDetector.test.ts`)
- [x] Planning prompts detected correctly (design, plan, architecture, should we)
- [x] Execution prompts detected correctly (implement, add, create, build)
- [x] Exploration prompts detected correctly (what if, could we, brainstorm)
- [x] Debugging prompts detected correctly (bug, error, fix, crash, not working)
- [x] Empty/ambiguous prompts default to execution

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added automatic task type detection that identifies user requests as planning, execution, exploration, or debugging tasks for improved contextual responses.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/lib/infrastructure/services/TaskTypeDetector.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Detect the task type from a user prompt.
+   */
+  detect(prompt: string): ITaskTypeResult {
+    if (!prompt || prompt.trim().length === 0) {
+      return { taskType: 'execution', confidence: 0, signals: [] };
+    }
+
+    const normalised = prompt.toLowerCase();
+
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply src/lib/infrastructure/services/TaskTypeDetector.ts]: Applied. Normalization now strips punctuation with `.replace(/[^\w\s]+/g, ' ')` so signals like "stack trace" match inputs like "stack-trace". Also collapses multiple spaces afterward.

File: src/lib/infrastructure/services/TaskTypeDetector.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        bestScore = score;
+        bestType = taskType;
+        bestSignals = matched;
+      }
+    }
+
+    // Compute confidence as proportion of winning score to total
+    const confidence = totalScore > 0
+      ? Math.min(bestScore / Math.max(totalScore, bestScore + 1), 1.0)
+      : 0;
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply src/lib/infrastructure/services/TaskTypeDetector.ts]: Good catch. Replaced the dampened formula with a dominance ratio: `bestScore / totalScore` — the fraction of all matched signal weight that belongs to the winning type. This avoids the problem with `bestScore / bestPossibleScore` where matching 2 out of 17 possible signals yields 0.118 (below the 0.15 threshold), even when the match is unambiguous. The dominance ratio correctly returns 1.0 when only one type matches and scales down as competing types also match.
